### PR TITLE
Initial commit of host metrics memory scraper

### PR DESCRIPTION
**Pull request by james-bebbington**
**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector/issues/847

**Description:**
Added memory scraper to the hostmetricsreceiver which uses gopsutil to collect memory used metric.

**Exported Metric on Windows:**
image

### CODE REVIEW

File: receiver/hostmetricsreceiver/hostmetrics_receiver_test.go [hunk 1]
```diff
@@ -55,30 +58,26 @@ func TestGatherMetrics_EndToEnd(t *testing.T) {
...
-	internal.AssertDescriptorEqual(t, cpuscraper.MetricCPUSecondsDescriptor, hostCPUTimeMetric.MetricDescriptor())
-	assert.GreaterOrEqual(t, hostCPUTimeMetric.Int64DataPoints().Len(), runtime.NumCPU()*4)
-	internal.AssertInt64MetricLabelExists(t, hostCPUTimeMetric, 0, cpuscraper.CPULabel)
-	internal.AssertInt64MetricLabelHasValue(t, hostCPUTimeMetric, 0, cpuscraper.StateLabel, cpuscraper.UserStateLabelValue)
-	internal.AssertInt64MetricLabelHasValue(t, hostCPUTimeMetric, 1, cpuscraper.StateLabel, cpuscraper.SystemStateLabelValue)
-	internal.AssertInt64MetricLabelHasValue(t, hostCPUTimeMetric, 2, cpuscraper.StateLabel, cpuscraper.IdleStateLabelValue)
-	internal.AssertInt64MetricLabelHasValue(t, hostCPUTimeMetric, 3, cpuscraper.StateLabel, cpuscraper.InterruptStateLabelValue)
+	// expected metric names
+	assert.Contains(t, metricNames, "host/cpu/time")
+	assert.Contains(t, metricNames, "host/memory/used")
```
Comment (james-bebbington) [inline receiver/hostmetricsreceiver/hostmetrics_receiver_test.go:82]: I simplified the e2e test quite a lot so it now only validates that we get all of the metrics expected, but doesn't validate the content of these metrics (which can be validated by tests in the scrapers). This means we don't replicate nearly identical validation logic for the scrapers and receivers, and also means we don't need to export constant values from the scrapers just to be used by this test (in a following PR I'll unexport the cpu metric constants). But it does make the e2e test a bit less thorough.

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+// Close
+func (c *Scraper) Close(ctx context.Context) error {
+	c.cancel()
+	return nil
+}
+
+func (c *Scraper) scrapeMetrics(ctx context.Context) {
+	ctx, span := trace.StartSpan(ctx, "memoryscraper.scrapeMetrics")
+	defer span.End()
```
Comment (james-bebbington) [inline receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper.go:73]: I was just looking around the repo a bit - should I be using `StartMetricsReceiveOp/EndMetricsReceiveOp` here instead of a custom trace?

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+// Factory is the Factory for scraper.
+type Factory struct {
+}
+
+// Type gets the type of the scraper config created by this Factory.
+func (f *Factory) Type() string {
+	return TypeStr
+}
```
Comment (bogdandrutu) [inline receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go:40]: This does not seem to be used. Do we need it?

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+// CreateDefaultConfig creates the default configuration for the Scraper.
+func (f *Factory) CreateDefaultConfig() internal.Config {
+	return &Config{}
+}
```
Comment (bogdandrutu) [inline receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go:45]: Would be good to provide some useful default for this so user can just write the scraper name in the config. Probably same comment applies to cpu, maybe do it in a separate PR?

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper.go (hunk 2 above)
Comment (bogdandrutu) [reply receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper.go:73]: Would be better I think.

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go (hunk 3 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go:40]: Hmm you're right, this isn't used. It's in the interface definition though - I'll remove it in a followup PR since it'll affect the cpu scraper too

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go (hunk 4 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/memoryscraper/factory.go:45]: The cpu scraper defines a useful default. The memory scraper doesn't have any config so this is just an empty struct, but the user does only need to write the scraper name in the config (see https://github.com/open-telemetry/opentelemetry-collector/pull/911/files#diff-a9325b7985e2499b8d97815fdbda4a26R11)

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper.go (hunk 2 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper.go:73]: Had a bit of a deeper look at this. From what I understand, `obsreport` is a helper package for generating standard traces & metrics for collector components, and currently uses the OpenCensus libraries for this.

A couple of questions if I want to use this here:

- I'm not sure what to provide for the "transport" attribute value (third parameter of `StartMetricsReceiveOp`). These metrics are scraped from the host rather than being "received" as such so it doesn't seem applicable - is it okay to pass an empty string?
- `obsreport.CountMetricPoints` is only implemented for the OC proto format. It probably needs a second implementation for the in-memory data structure rather than doing a transformation in the code (should be relatively straightforward to do, but probably better done in a separate PR)
